genienlp/models/transformer_seq2seq.py

```diff
@@ -159,3 +162,84 @@ def generate(self,
                                         )
 
         return generated
+
+
+    def confidence(self, batch, predictions, mc_dropout=False, mc_dropout_num=0) -> List[ConfidenceOutput]:
+        """
+        predictions: Tensor of shape (batch_size, output_length)
+        mc_dropout: if True, will activate dropout layers
+        mc_droput_num: number of Monte Carlo samples used for the MC Dropout method
+        """
+        if mc_dropout:
+            assert mc_dropout_num > 0, 'MC Dropout is enabled, but mc_droput_num is 0'
+        else:
+            assert mc_dropout_num == 0, 'MC Dropout is disabled, but mc_droput_num is not 0'
+        batch_size = predictions.shape[0]
+        repetition_factor = batch_size//batch.context.value.shape[0]
+        input_ids = batch.context.value.repeat_interleave(repetition_factor, dim=0) # repeat to account for multiple predictions per input
+
+        prediction_lengths = self.get_length(predictions)
+
+        pad_token_id = self.numericalizer._tokenizer.pad_token_id
+        attention_mask = self.model._prepare_attention_mask_for_generation(input_ids=input_ids, pad_token_id=pad_token_id, eos_token_id=self.numericalizer._tokenizer.eos_token_id)
+        truncated_predictions = predictions[:, 1:] # remove the BOS token since it is not actually being generated
+
+        batch_nodrop_logits = []
+        batch_nodrop_probs = []
+        batch_nodrop_entropies = []
+        outputs = self.model(input_ids=input_ids, decoder_input_ids=predictions, attention_mask=attention_mask, return_dict=True, use_cache=False)
+        nodrop_logits = outputs.logits[:, :-1, :] # remove the last probability distribution which is for the token after EOS
+        for i in range(batch_size):
+            batch_nodrop_logits.append(nodrop_logits[i].gather(dim=1, index=truncated_predictions[i].view(-1, 1)).view(-1)[:prediction_lengths[i]])
+            probs = torch.softmax(nodrop_logits[i], dim=1)
+            batch_nodrop_probs.append(probs.gather(dim=1, index=truncated_predictions[i].view(-1, 1)).view(-1)[:prediction_lengths[i]])
+            batch_nodrop_entropies.append(-torch.sum(torch.log(probs)*probs, dim=1)[:prediction_lengths[i]])
+        
+        # activate dropout layers
+        should_revert_to_eval = True
+        if mc_dropout:
```

Comment:
`should_revert_to_eval` and `mc_dropout` are different things. It is possible that the caller has already put the model into `train()` mode, so after tha call is done, it should still be in `train()` mode.

Reply:
If there is no mc_dropout, you don't call .train(), so you also don't need to call .eval() (that's why you check for both later).
I think it would be more readable as:
```
should_revert_to_eval = False
if mc_dropout not self.training:
    self.train()
    should_rever_to_eval = True
```